aidial_adapter_vertexai/chat/tools.py

```diff
@@ -0,0 +1,209 @@
+from typing import Dict, List, Literal, Self, Tuple, assert_never
+
+from aidial_sdk.chat_completion import (
+    Function,
+    FunctionChoice,
+    Message,
+    Role,
+    ToolChoice,
+)
+from aidial_sdk.chat_completion.request import AzureChatCompletionRequest
+from pydantic import BaseModel
+from vertexai.preview.generative_models import (
+    FunctionDeclaration as GeminiFunction,
+)
+from vertexai.preview.generative_models import Tool as GeminiTool
+from vertexai.preview.generative_models import ToolConfig as GeminiToolConfig
+
+from aidial_adapter_vertexai.chat.errors import ValidationError
+
+FunctionCallingConfig = GeminiToolConfig.FunctionCallingConfig
+
+
+class ToolsConfig(BaseModel):
+    functions: List[Function]
+    """
+    List of functions/tools.
+    """
+
+    required: bool
+    """
+    True forces the model to call one of the available functions.
+    False allows the model to pick between generating a message or
+    calling one or more tools/functions.
+    """
+
+    tool_ids: Dict[str, str] | None
+    """
+    Mapping from tool call IDs to corresponding tool names.
+    None means that functions are used, not tools.
+    """
+
+    @property
+    def is_tool(self) -> bool:
+        return self.tool_ids is not None
+
+    def not_supported(self) -> None:
+        if self.functions:
+            if self.is_tool:
+                raise ValidationError("The tools aren't supported")
+            else:
+                raise ValidationError("The functions aren't supported")
+
+    def create_fresh_tool_call_id(self, tool_name: str) -> str:
+        if self.tool_ids is None:
+            raise ValidationError("Function are used, but requested tool id")
+
+        idx = 1
+        while True:
+            id = f"{tool_name}_{idx}"
+            if id not in self.tool_ids:
+                self.tool_ids[id] = tool_name
+                return id
+            idx += 1
+
+    def get_tool_name(self, tool_call_id: str) -> str:
+        if self.tool_ids is None:
+            raise ValidationError("Function are used, but requested tool name")
+
+        tool_name = self.tool_ids.get(tool_call_id)
+        if tool_name is None:
+            raise ValidationError(f"Tool call ID not found: {self.tool_ids}")
+        return tool_name
+
+    @staticmethod
+    def filter_functions(
+        function_call: Literal["auto", "none", "required"] | FunctionChoice,
+        functions: List[Function],
+    ) -> Tuple[bool, List[Function]]:
+        match function_call:
+            case "none":
+                return False, []
+            case "auto":
+                return False, functions
+            case "required":
```

Comment:
Where does "required" come from?

Reply:
It's from OpenAI docs: https://platform.openai.com/docs/api-reference/chat/create#chat-create-tool_choice
But it seems like Azure OpenAI doesn't support it. I've removed it,